contents/go/Go.md

```diff
@@ -0,0 +1,273 @@
+
+Specifically, Go relies on channels to implement message passing. Channels are typed conduits that allows goroutines to communicate with each other by sending and receiving messages. Before using a channel of a specific type, we must declare and `make` it:
+
+```go
+var c chan int
+c = make(chan int)
+
+//alternatively
+c := make(chan int)
+```
+
+Channels can transmit data of any type; thus, creating a channel that transmits channels (i.e. a `chan chan`) is theoretically possible and may even be [useful](http://tleyden.github.io/blog/2013/11/23/understanding-chan-chans-in-go/).
+
+Goroutines send and receive messages through a channel using the `<-` operator.
+
+```go
+ch <- v   // Send v to channel ch.
+v := <-ch // Receive from ch, and
+	  // assign value to v.
+```
+One useful way to think about sending and receiving data with the `<-` operator is that the data moves in the direction of the arrow.
+
+Channels can be used to synchronize execution across goroutines, since receivers block until they receive data, while senders block [until the receiver or buffer receives data](https://golang.org/doc/effective_go.html#channels). In the code example below, the `main` goroutine waits until it receives a message from the `worker` goroutine that it is done before terminating.
+
+```go
+func worker(done chan bool) {
+	fmt.Print("Working...")
+	time.Sleep(time.Second)
+	fmt.Println("Done working")
+
+	done <- true
+}
+
+func main() {
+	done := make(chan bool)
+	go worker(done)
+
+	<-done
+	fmt.Println("Returned from work")
+}
+```
+
+If you are interested in delving deeper into using Go's concurrency features extensively, Google developers have put out video presentations on Go's [basic](https://www.youtube.com/watch?v=f6kdp27TYZs) and [advanced concurrency patterns](https://www.youtube.com/watch?v=QDDwwePbDtw). This [code walkthrough](https://golang.org/doc/codewalk/sharemem/) provides an annotated example of how Go's memory-sharing principles can be applied in practice.
+
+### `defer`
+As opposed to traditional control flow mechanisms such as `if`, `for` and `switch`, which execute functions immediately, Go's `defer` keyword pushes a function call to a list, and only executes all functions on the list after the surrounding function returns. In the following example, `defer` adds two print functions to the stack of deferred functions. After `foo` finishes executing, the deferred functions are executed in last-in-first-out order.
+
+```go
+func foo() {
+	defer fmt.Println("This gets printed third")
+	defer fmt.Println("This gets printed second")
+	fmt.Println("This gets printed first")
+}
+```
+
+`defer` is frequently used for clean-up actions, such as to [close files](https://gobyexample.com/defer). Deferred functions run on [panicking goroutines](#error-handling) as well, which makes them useful for recovering from `panic`.
```

Comment:
The mention of _panicking goroutines_  hurts the self-contained quality of this article because the reader has to go to another page before continuing to read this article. Give a quick definition (or avoid altogether)?

Reply:
Decided to move this section further down in the list of Go features. The hope is that users unfamiliar with Go will have read the previous section on `panic`, which will teach them about what a panicking goroutine is, before they reach this section.